File: Airship/AirshipCore/Source/common/UAAppStateTracker.m
/* Copyright Airship and Contributors */

#import "UAAppStateTracker.h"
#import "UAAppStateTrackerAdapter+Internal.h"
#import "UAUIKitStateTrackerAdapter+Internal.h"

NSNotificationName const UAApplicationDidFinishLaunchingNotification = @"com.urbanairship.applicaiton_did_finish_launching";
NSNotificationName const UAApplicationDidBecomeActiveNotification = @"com.urbanairship.application_did_become_active";
NSNotificationName const UAApplicationWillEnterForegroundNotification = @"com.urbanairship.application_will_enter_foreground";
NSNotificationName const UAApplicationDidEnterBackgroundNotification = @"com.urbanairship.application_did_enter_background";
NSNotificationName const UAApplicationWillResignActiveNotification = @"com.urbanairship.application_will_resign_active";
NSNotificationName const UAApplicationWillTerminateNotification = @"com.urbanairship.application_will_terminate";
NSNotificationName const UAApplicationDidTransitionToBackground = @"com.urbanairship.application_did_transition_to_background";
NSNotificationName const UAApplicationDidTransitionToForeground = @"com.urbanairship.application_did_transition_to_foreground";

NSString *const UAApplicationLaunchOptionsRemoteNotificationKey = @"com.urbanairship.application_launch_options_remote_notification";

@interface UAAppStateTracker() <UAAppStateTrackerDelegate>
@property (nonatomic, strong) id<UAAppStateTrackerAdapter> adapter;
@property (nonatomic, strong) NSNotificationCenter *notificationCenter;
@property (nonatomic, assign) BOOL isForegrounded;
@end

@implementation UAAppStateTracker

static UAAppStateTracker *shared_;

+ (void)load {
    shared_ = [[UAAppStateTracker alloc] init];
}

+ (instancetype)shared {
    return shared_;
}

- (instancetype)init {
    self = [super init];

    if (self) {
        self.notificationCenter = [NSNotificationCenter defaultCenter];
        self.adapter = [UAUIKitStateTrackerAdapter adapter];
        self.adapter.stateTrackerDelegate = self;
    }

    return self;
}

- (void)applicationDidFinishLaunching:(NSDictionary *)remoteNotification {
    NSDictionary *userInfo = remoteNotification == nil ? nil : @{ UAApplicationLaunchOptionsRemoteNotificationKey : remoteNotification };
    [self.notificationCenter postNotificationName:UAApplicationDidFinishLaunchingNotification
                                           object:nil
                                         userInfo:userInfo];
}

- (void)applicationDidBecomeActive {
    [self.notificationCenter postNotificationName:UAApplicationDidBecomeActiveNotification object:nil];

    if (!self.isForegrounded) {
        self.isForegrounded = YES;
        [self.notificationCenter postNotificationName:UAApplicationDidTransitionToForeground object:nil];
    }
}

- (void)applicationWillEnterForeground {
    [self.notificationCenter postNotificationName:UAApplicationWillEnterForegroundNotification object:nil];
}

- (void)applicationDidEnterBackground {
    [self.notificationCenter postNotificationName:UAApplicationDidEnterBackgroundNotification object:nil];

    if (self.isForegrounded) {
        self.isForegrounded = NO;
        [self.notificationCenter postNotificationName:UAApplicationDidTransitionToBackground object:nil];
    }
}

- (void)applicationWillResignActive {
    [self.notificationCenter postNotificationName:UAApplicationWillResignActiveNotification object:nil];
}

- (void)applicationWillTerminate {
    [self.notificationCenter postNotificationName:UAApplicationWillTerminateNotification object:nil];
}

@end

File: Airship/AirshipCore/Source/UAAppStateTracker.swift
/* Copyright Airship and Contributors */


@objc
open class UAAppStateTracker: NSObject, UAAppStateTrackerDelegate {

    @objc
    public static let didBecomeActiveNotification = NSNotification.Name("com.urbanairship.application_did_become_active")

    @objc
    public static let willEnterForegroundNotification = NSNotification.Name("com.urbanairship.application_will_enter_foreground")

    @objc
    public static let didEnterBackgroundNotification = NSNotification.Name("com.urbanairship.application_did_enter_background")

    @objc
    public static let willResignActiveNotification = NSNotification.Name("com.urbanairship.application_will_resign_active")

    @objc
    public static let willTerminateNotification = NSNotification.Name("com.urbanairship.application_will_terminate")

    @objc
    public static let didTransitionToBackground = NSNotification.Name("com.urbanairship.application_did_transition_to_background")

    @objc
    public static let didTransitionToForeground = NSNotification.Name("com.urbanairship.application_did_transition_to_foreground")

    @objc
    public static let shared: UAAppStateTracker = UAAppStateTracker()

    private var adapter: UAAppStateTrackerAdapter?
    private let notificationCenter: NotificationCenter
    private var isForegrounded: Bool

    @objc
    open var state: UAApplicationState {
        get {
            return adapter?.state ?? .background
        }
    }

    @objc
    public init(notificationCenter: NotificationCenter, adapter: UAAppStateTrackerAdapter?) {
        self.notificationCenter = notificationCenter
        self.adapter = adapter
        self.isForegrounded = adapter?.state == .active
        super.init()
        self.adapter?.stateTrackerDelegate = self
    }

    public override convenience init() {
        self.init(notificationCenter: NotificationCenter.default, adapter: UAUIKitStateTrackerAdapter())
    }

    @objc
    public func applicationDidBecomeActive() {
        notificationCenter.post(name: UAAppStateTracker.didBecomeActiveNotification, object: nil)

        if !isForegrounded {
            isForegrounded = true
            notificationCenter.post(name: UAAppStateTracker.didTransitionToForeground, object: nil)
        }
    }

    @objc
    public func applicationWillEnterForeground() {
        notificationCenter.post(name: UAAppStateTracker.willEnterForegroundNotification, object: nil)
    }

    @objc
    public func applicationDidEnterBackground() {
        notificationCenter.post(name: UAAppStateTracker.didEnterBackgroundNotification, object: nil)

        if isForegrounded {
            isForegrounded = false
            notificationCenter.post(name: UAAppStateTracker.didTransitionToBackground, object: nil)
        }
    }

    @objc
    public func applicationWillResignActive() {
        notificationCenter.post(name: UAAppStateTracker.willResignActiveNotification, object: nil)
    }

    @objc
    public func applicationWillTerminate() {
        notificationCenter.post(name: UAAppStateTracker.willTerminateNotification, object: nil)
    }
}
